cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/CaReconciler.java

```diff
@@ -0,0 +1,515 @@
+        this.featureGates = config.featureGates();
+
+        this.deploymentOperator = supplier.deploymentOperations;
+        this.stsOperator = supplier.stsOperations;
+        this.strimziPodSetOperator = supplier.strimziPodSetOperator;
+        this.secretOperator = supplier.secretOperations;
+        this.podOperator = supplier.podOperations;
+
+        this.adminClientProvider = supplier.adminClientProvider;
+        this.zookeeperLeaderFinder = supplier.zookeeperLeaderFinder;
+        this.certManager = certManager;
+        this.passwordGenerator = passwordGenerator;
+
+        // Extract required information from the Kafka CR
+        this.maintenanceWindows = kafkaCr.getSpec().getMaintenanceTimeWindows();
+        this.ownerRef = new OwnerReferenceBuilder()
+                .withApiVersion(kafkaCr.getApiVersion())
+                .withKind(kafkaCr.getKind())
+                .withName(kafkaCr.getMetadata().getName())
+                .withUid(kafkaCr.getMetadata().getUid())
+                .withBlockOwnerDeletion(false)
+                .withController(false)
+                .build();
+        this.clusterCaConfig = kafkaCr.getSpec().getClusterCa();
+        this.clientsCaConfig = kafkaCr.getSpec().getClientsCa();
+        this.caLabels = Labels
+                .generateDefaultLabels(kafkaCr, Labels.APPLICATION_NAME, AbstractModel.STRIMZI_CLUSTER_OPERATOR_NAME)
+                .toMap();
+        this.clusterOperatorSecretLabels = Labels.fromResource(kafkaCr)
+                .withStrimziKind(reconciliation.kind())
+                .withStrimziCluster(reconciliation.name())
+                .withKubernetesName(Labels.APPLICATION_NAME)
+                .withKubernetesInstance(reconciliation.name())
+                .withKubernetesPartOf(reconciliation.name())
+                .withKubernetesManagedBy(AbstractModel.STRIMZI_CLUSTER_OPERATOR_NAME);
+        this.clusterCaCertLabels = clusterCaCertLabels(kafkaCr);
+        this.clusterCaCertAnnotations = clusterCaCertAnnotations(kafkaCr);
+    }
+
+    /**
+     * Utility method to extract the template labels from the Kafka CR.
+     *
+     * @param kafkaCr   Kafka CR
+     *
+     * @return  Map with the labels from the Kafka CR or empty map if the template is not set
+     */
+    private static Map<String, String> clusterCaCertLabels(Kafka kafkaCr)    {
+        if (kafkaCr.getSpec().getKafka() != null
+                && kafkaCr.getSpec().getKafka().getTemplate() != null
+                && kafkaCr.getSpec().getKafka().getTemplate().getClusterCaCert() != null
+                && kafkaCr.getSpec().getKafka().getTemplate().getClusterCaCert().getMetadata() != null
+                && kafkaCr.getSpec().getKafka().getTemplate().getClusterCaCert().getMetadata().getLabels() != null) {
+            return kafkaCr.getSpec().getKafka().getTemplate().getClusterCaCert().getMetadata().getLabels();
+        } else {
+            return Map.of();
+        }
+    }
```

Comment:
Maybe this method could be something like this to avoid all this explicit `null` checks?

```java
return Optional.ofNullable(kafkaCr.getSpec().getKafka())
		.map(k -> k.getTemplate())
		.map(t -> t.getClusterCaCert()
		.map(r -> r.getMetadata())
		.map(m -> r.getLabels())
		.orElse(Map.of());
```

Reply:
TBH, I think the `if` above is way more readable and easier to understand what it actually does.

Reply:
I can see your point, that's the good old way but too verbose. Btw no need to change. It was just a suggestion.

Reply:
Thanks.